### ISSUE TITLE
Confusing behaviour with prechipped rasters and intersection/union datasets

### ISSUE THREAD

**Issue by Spiruel** [microsoft/torchgeo#1539]
### Description

I have prechipped geotiffs in a folder, and I'm combining them with a shapefile of field boundaries. 

When getting the intersection of the raster and vector data, I am getting repeated chips in my dataset. Using a union of the raster and vector data seems to avoid these repeated chips. However running a simple unet training run on this uniondataset causes the kernel to inexplicably crash.

### Steps to reproduce

```
class HLS_S30(RasterDataset):
    filename_glob = "*.tif"
    filename_regex = r"""
    ^\d+_\d+_(?P<date>\d{8})T?.tif$
    """
    date_format = "%Y%m%d"
    
    all_bands = [
            "B01",
            "B02",
            "B03",
            "B04",
            "B05",
            "B06",
            "B07",
            "B08",
            "B8A",
            "B09",
            "B10",
            "B11",
            "B12",
        ]
    rgb_bands = ["B04", "B03", "B02"]

    def __init__(
            self,
            root,
            crs = None,
            res = None,
            bands = None,
            transforms= None,
            cache = True,
        ) -> None:
        bands = bands or self.all_bands

        super().__init__(root, crs=crs, res=res, bands=bands, transforms=transforms, cache=cache)

field_parcels = VectorDataset(root='cc_data', crs=None, res=0.0001, transforms=None)
image_folder = 'HLS_S30/4'
ds_inputs = HLS_S30(root=image_folder, transforms=None)

ds = ds_inputs & field_parcels
len(ds)
```

len(ds) = 534 with `ds_inputs & field_parcels` and len(ds) = 181 with `ds_inputs | field_parcels`. The expected number is 178.

Training code to reproduce kernel crashing:

```
import torch
import torch.nn as nn

class UNet(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(UNet, self).__init__()
        
        # Encoder
        self.encoder = nn.Sequential(
            nn.Conv2d(in_channels, 64, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),
            nn.Conv2d(64, 64, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2, stride=2)
        )
        
        # Decoder
        self.decoder = nn.Sequential(
            nn.Conv2d(64, 64, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),
            nn.Conv2d(64, 64, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),
            nn.ConvTranspose2d(64, out_channels, kernel_size=2, stride=2)
        )
    
    def forward(self, x):
        # Encoder
        enc_out = self.encoder(x)
        
        # Decoder
        dec_out = self.decoder(enc_out)
        
        return dec_out

# Create an instance of the UNet model
in_channels = 13  # Number of input channels
out_channels = 1  # Number of output channels (for segmentation mask)
model = UNet(in_channels, out_channels).cuda()

train_batch_sampler = PreChippedGeoSampler(ds)#, roi=train_roi)
val_batch_sampler = PreChippedGeoSampler(ds)#, roi=val_roi)

train_dl = torch.utils.data.DataLoader(ds, batch_size=16, sampler=train_batch_sampler, collate_fn=stack_samples)

# Define loss function and optimizer
criterion = nn.BCEWithLogitsLoss()  # Binary cross-entropy loss for binary segmentation
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# Training loop
num_epochs = 10

for epoch in range(num_epochs):
    model.train()  # Set the model to training mode
    
    for sample in tqdm(train_dl):
        input_features = transform(sample["image"]).cuda()
        labels = transform(sample["mask"]).float().squeeze().cuda()
        # Forward pass
        outputs = model(input_features).squeeze()

        # Calculate loss
        loss = criterion(outputs, labels)

        # Backpropagation and optimisation
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

print("Training complete!")
```

### Version

0.4

**Comment by adamjstewart** [comment]
> When getting the intersection of the raster and vector data, I am getting repeated chips in my dataset.

Intersection will compute all possible intersections. So if you have images A and B and masks C and D and all 4 have at least some spatiotemporal overlap, the resulting IntersectionDataset will have length 4, not 2.

> Using a union of the raster and vector data seems to avoid these repeated chips.

Depending on what you're trying to do, union doesn't make a lot of sense. Union corresponds to "or", so if you have images A and B and masks C and D you will have a dataset of length 4 where one of [A, B, C, D] will randomly be selected. It's designed for when you have images A and B and images C and D and you want images A/B/C/D.

If your dataset is pre-chipped, you may be better off creating a NonGeoDataset. This will work better in the case that your images have partial overlap. Note that touching edges count as overlap, even if the area of overlap is 0.
